extension/src/countminsketch.rs

```diff
@@ -0,0 +1,275 @@
+    pg_type,
+    raw::bytea,
+    ron_inout_funcs,
+};
+
+#[pg_schema]
+pub mod toolkit_experimental {
+    use super::*;
+
+    pg_type! {
+        #[derive(Debug)]
+        struct CountMinSketch<'input> {
+            width: u32,
+            depth: u32,
+            counters: [i64; self.width * self.depth],
+        }
+    }
+
+    impl CountMinSketch<'_> {
+        fn new(width: u32, depth: u32, counters: Vec<i64>) -> Self {
+            let counters_arr = counters.try_into().unwrap();
+            unsafe {
+                flatten!(CountMinSketch {
+                    width,
+                    depth,
+                    counters: counters_arr,
+                })
+            }
+        }
+
+        pub fn to_internal_countminsketch(&self) -> CountMinSketchInternal {
+            let depth: u64 = self.depth.into();
+            let hashfuncs = (1..=depth).map(CountMinHashFn::with_key).collect();
+
+            let mut counters: Vec<Vec<i64>> = Vec::with_capacity(self.depth as usize);
+            let row_width = self.width as usize;
+            for row in 0..self.depth {
+                let row_start = (row * self.width) as usize;
+                counters.push(
+                    self.counters
+                        .iter()
+                        .skip(row_start)
+                        .take(row_width)
+                        .collect(),
+                );
+            }
+
+            CountMinSketchInternal::new(
+                self.width as usize,
+                self.depth as usize,
+                hashfuncs,
+                counters,
+            )
+        }
+
+        pub fn from_internal_countminsketch(sketch: &CountMinSketchInternal) -> Self {
+            CountMinSketch::new(
+                sketch.width().try_into().unwrap(),
+                sketch.depth().try_into().unwrap(),
+                sketch.counters().iter().flatten().cloned().collect(),
+            )
+        }
+    }
+
+    ron_inout_funcs!(CountMinSketch);
+}
+
+use toolkit_experimental::CountMinSketch;
+
+#[aggregate]
+impl toolkit_experimental::count_min_sketch {
+    type State = CountMinSketchInternal;
+
+    fn transition(
+        state: Option<State>,
+        #[sql_type("text")] value: Option<String>,
+        #[sql_type("float")] error: f64,
+        #[sql_type("float")] probability: f64,
+    ) -> Option<State> {
+        let value = match value {
+            None => return state,
+            Some(value) => value,
+        };
+
+        let mut state = match state {
+            None => CountMinSketchInternal::with_prob(error, probability),
+            Some(state) => state,
+        };
+
+        state.add_value(value);
+        Some(state)
+    }
+
+    fn finally(state: Option<&mut State>) -> Option<CountMinSketch<'static>> {
+        let state = match state {
+            None => return None,
+            Some(state) => CountMinSketch::from_internal_countminsketch(state),
+        };
+
+        Some(state)
```

Comment:
Can this function just be `state.map(CountMinSketch::from_internal_countminsketch)`